LoRaEngine/LoraKeysManagerFacade/DeviceGetter.cs

```diff
@@ -19,126 +19,82 @@ namespace LoraKeysManagerFacade
-                        if (redisConnectionString == null)
-                        {
-                            string errorMsg = "Missing RedisConnectionString in settings";
-                            throw new Exception(errorMsg);
-                        }
-
-                        registryManager = RegistryManager.CreateFromConnectionString(connectionString);
-
-                        var redis = ConnectionMultiplexer.Connect(redisConnectionString);
-                        redisCache = redis.GetDatabase();
-                    }
-                }
+                var results = await GetDeviceList(devEUI, gatewayId, devNonce, devAddr, context);
+                string json = JsonConvert.SerializeObject(results);
+                return new OkObjectResult(json);
+            }
+            catch (DeviceNonceUsedException)
+            {
+                return new BadRequestObjectResult("UsedDevNonce");
             }
+        }
 
-            List<IoTHubDeviceInfo> results = new List<IoTHubDeviceInfo>();
+        public static async Task<List<IoTHubDeviceInfo>> GetDeviceList(string devEUI, string gatewayId, string devNonce, string devAddr, ExecutionContext context)
+        {
+            var results = new List<IoTHubDeviceInfo>();
+            var registryManager = LoRaRegistryManager.GetCurrentInstance(context.FunctionAppDirectory);
 
-            // OTAA join
             if (devEUI != null)
             {
+                // OTAA join
                 string cacheKey = devEUI + devNonce;
-
-                try
+                using (var deviceCache = LoRaDeviceCache.Create(context, devEUI, gatewayId, cacheKey))
                 {
-                    if (redisCache.LockTake(cacheKey + "joinlock", gatewayId, new TimeSpan(0, 0, 10)))
+                    if (deviceCache.TryToLock(cacheKey + "joinlock"))
                     {
-                        // check if we already got the same devEUI and devNonce it can be a reaply attack or a multigateway setup recieving the same join.We are rfusing all other than the first one.
-                        string cachedDevNonce = redisCache.StringGet(cacheKey, CommandFlags.DemandMaster);
-
-                        if (!string.IsNullOrEmpty(cachedDevNonce))
+                        if (deviceCache.TryGetValue(out _))
                         {
-                            return (ActionResult)new BadRequestObjectResult("UsedDevNonce");
+                            throw new DeviceNonceUsedException();
                         }
 
-                        redisCache.StringSet(cacheKey, devNonce, new TimeSpan(0, 1, 0), When.Always, CommandFlags.DemandMaster);
+                        deviceCache.SetValue(devNonce, TimeSpan.FromMinutes(1));
 
-                        IoTHubDeviceInfo iotHubDeviceInfo = new IoTHubDeviceInfo();
                         var device = await registryManager.GetDeviceAsync(devEUI);
```

Comment:
Here is something that we have never implemented but as a security measuere we should check that the lengh is exaclty the one of DevEUI (I belive 16) and is a valid hex format. In case the function key get compromise you can get only the lora devices out and not all the other iot hub devices. You don't neet to add it to this PR but would make sense to have it in here at some point